Experiment on converting from one format to the other

GabrielSimonetto opened the pull request:
1) Using `serde-transcode` in order to convert one format into the other:

The first commit of this PR focuses on trying to use `serde-transcode` as the solution to `BioSerDe` proposal. This however does not currently work, because we cannot give the serde mechanism a hint in which type we are expecting:

```rust
// Failing serde-transcode example:
// From json representation to noodles-bed representation
let input = r#"{"chrom":"sq0","start":8,"end":13}"#;
let mut deserializer = serde_json::Deserializer::from_str(input);
let mut serializer = noodles_bed::RecordSerializer::new();
serde_transcode::transcode(&mut deserializer, &mut serializer).unwrap();
```

Notice that the only information inside this is the Serializer and Deserializer of the target formats, whereas our currently code only works when we explicitly use the `SerdeRecordWrapper` via our auxiliar functions.

One problem with not using transcode would be the use of intermediary memory, aswell as a reduction in the simplicity of the code:

`The [serde-transcode](https://github.com/sfackler/serde-transcode) crate provides functionality to "transcode" from an arbitrary Serde Deserializer to an arbitrary Serde Serializer without needing to collect the entire input into an intermediate form in memory.`

But it seems to me that in order to use transcode, we cannot give the serde mechanism a hint in which type we are expecting, if we do that, we will need to go back in our decision of using `serde-with` in order to use the `FromStr` and `Display` already implemented in the noodles crate to accelerate the experiment.

Another problem with **using** transcode, is that we need to have access to the `Serializer` and `Deserializer` of the format we are dealing with, however, some crates may choose to keep those objects private. That seems to be the case for the `rust-csv` crate for example, if I didn't mess up my demo earlier, it could be that the only formats that make this choice will not be the ones where it makes sense to use in bioserde, especially since WE are the ones implementing serde for at least the noodles formats, so they would always be public for our own sake hehe.

2) Trying a manual approach

In order to weight if we really have to eat this bullet, I've tried to explore how would we do this functionality manually on the subsequential commits of this branch, currently it isn't going great:

First I've tried to create single functions for each format:

```rust
    pub fn noodles_bed_to_json_bed(input: &str) -> String {
        let record: Record<3> = noodles_bed::record_from_str(input).unwrap();

        // TODO: start treating errors
        serde_json::to_string(&record).unwrap()
    }
```

That doesn't sound like it's correct, since we would need one function for each Record<N>.

Then I've tried to make a more generic function with an enum of `SupportedFormats`, which initially, could not generalize, since it would expect many Record<N> types in the same function, which rust does not allow

```rust
    pub fn convert_to_format(input: &str, from: SupportedFormat, to: SupportedFormat) -> String {
        let record = match from {
            SupportedFormat::Record3 => noodles_bed::record_from_str::<Record<3>>(input).unwrap(),
            SupportedFormat::Record4 => noodles_bed::record_from_str::<Record<4>>(input).unwrap(),
            SupportedFormat::JsonBed3 => serde_json::from_str::<Record<3>>(input).unwrap(),
            SupportedFormat::JsonBed4 => serde_json::from_str::<Record<4>>(input).unwrap(),
        };

        match to {
            SupportedFormat::Record3 | SupportedFormat::Record4 => {
                noodles_bed::record_to_string(record).unwrap()
            }
            SupportedFormat::JsonBed3 | SupportedFormat::JsonBed4 => {
                serde_json::to_string(&record).unwrap()
            }
        }
    }
```

```
error[E0308]: `match` arms have incompatible types
  --> noodles-bed/src/main.rs:46:41
   |
44 |           let record = match from {
   |  ______________________-
45 | |             SupportedFormat::Record3 => noodles_bed::record_from_str::<Record<3>>(input).unwrap(),
   | |                                         --------------------------------------------------------- this is found to be of type `Record<3_u8>`
46 | |             SupportedFormat::Record4 => noodles_bed::record_from_str::<Record<4>>(input).unwrap(),
   | |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `3_u8`, found `4_u8`
47 | |             SupportedFormat::JsonBed3 => serde_json::from_str::<Record<3>>(input).unwrap(),
48 | |             SupportedFormat::JsonBed4 => serde_json::from_str::<Record<4>>(input).unwrap(),
49 | |         };
   | |_________- `match` arms have incompatible types
   |
   = note: expected struct `Record<3_u8>`
              found struct `Record<4_u8>`
```

Then, I've realized that we can merge the 2 approaches to have a function that does the whole serde process for each input, which works. (I will not print it here, since it's the current version) 

However it's really verbose, we need both a) better ways to fill the match patterns in a way that allows only valid conversions, and b) probably, use macros to create all the functions needed, (which raises exponentially with each added format), that however, sound like it could work well.

My current plan is to persue these enhancements and see if the code is good enough, if however you guys have questions or other ideas, please let me know (:

p.s.: the current code is kinda ugly and have some things I still have to check (such as, if a JsonBed3 and JsonBed4 are really needed), but I still wanted to submit the architecture idea.

GabrielSimonetto committed:
Failing example of serde-transcode

GabrielSimonetto committed:
wip: initial bioserde demo

GabrielSimonetto committed:
Introduce convert_to_format architecture

GabrielSimonetto committed:
Error: match arms have incompatible types

mmalenic commented on noodles-bed/src/main.rs:
```diff
@@ -1,6 +1,101 @@
+        JsonBed4,
+    }
+
+    pub fn convert_to_format(input: &str, from: SupportedFormat, to: SupportedFormat) -> String {
```
The compiler won't allow different match arm types here. You _could_ potentially merge the two match branches, for example:

```rust
match (from, to) {
    (SupportedFormat::Record3, SupportedFormat::Record3) => noodles_bed::record_to_string(noodles_bed::record_from_str::<Record<3>>(input).unwrap()).unwrap(),
    ...
}
```

Or use some sort of common trait and dynamically dispatch in a [`Box`](https://doc.rust-lang.org/std/boxed/struct.Box.html).

However, neither seem like their are the right way to go about it, and are still verbose.

mmalenic reviewed (commented):
Thanks @GabrielSimonetto! With regards to serde-transcode, I still think this is the way to go if possible. I feel like attempting to somehow unify `Record<N>` and `SerdeRecordWrapper` with how they are serialized/deserialized would resolve some of the complexity. 

Are you sure there is no way serde_with can help us? This could be unrelated, but does [`DeserializeFromStr`](https://docs.rs/serde_with/latest/serde_with/derive.DeserializeFromStr.html) or [`SerializeDisplay`](https://docs.rs/serde_with/latest/serde_with/derive.SerializeDisplay.html) help in any way?

Maybe the `T` bounds could be restricted in some way to differentiate between the JSON and BED implementations of `Serializer` and `Deserializer`? 🤔

GabrielSimonetto committed:
wip: unifying types on match

GabrielSimonetto committed:
Revert to serde-transcode approach to connecting serde pipelines

    Revert "wip: unifying types on match"
    Revert "Error: match arms have incompatible types"
    Revert "Introduce convert_to_format architecture"
    Revert "wip: initial bioserde demo"

GabrielSimonetto committed:
wip: remove serde_with from dependencies

- Go back to the serializer used on commit:
    0484112649c62e66bb6d1b879c66579a3822da71

- Change tests that pass with serde_with utilities,
    to use to_string() and from_string() directly,
    (which are functions that do not use type hints)

- TODO: Make tests pass
